### ISSUE TITLE
Proxy resolution of singleton types

### ISSUE THREAD

**Issue by gpauloski** [proxystore/proxystore#311]
### Describe the problem.

Resolving a proxy where the target object is `None` will raise an error indicating the target object could not be resolved because the proxy conflates `None` with being the default return value of `Store.get()` for when the object does not exists.

This is an easy fix: `StoreFactory.resolve` can pass a sentinel object as the default return value to `Store.get()`.

However, there's still another issue with proxying singleton types (`True`/`False`/`None`). Identity comparison of a proxy to a singleton will 1) always fail and 2) not cause the proxy to resolve. E.g., `proxy(lambda: True) is True` evaluates as false.

There are a few approaches we could take to this:
1) Not allow proxying singleton types.
    - Seems extreme to me and could lead to users having to write extra code to handle those cases.
2) Add a warning when a singleton type is proxied.
3) Do nothing and just document the caveat.

```python
import tempfile

from proxystore.connectors.file import FileConnector
from proxystore.proxy import is_resolved
from proxystore.store import Store

with tempfile.TemporaryDirectory() as tmp_dir:
    with Store('test', FileConnector(tmp_dir)) as store:
        # Storing/retrieving None is okay
        key = store.put(None)
        assert store.get(key) is None

        proxy = store.proxy(None)
        # Singleton comparision does not trigger resolve and this fails
        assert proxy is not None
        assert not is_resolved(proxy)
        # Also fails, but because the proxy thinks the object does not exist
        assert proxy == None
```

Issue found by @benclifford.

### How did you install ProxyStore?

```console
$ pip install proxystore
```

### ProxyStore Version

v0.5.0

### Python Version

3.11

### OS and Platform

x86 Linux

**Comment by benclifford** [comment]
my gut feeling from dealing with singleton serialization in a different situation is "don't try to make it work" (i.e. option 1 in the issue text). 

I think this reason to not do option 1 is weak 

> could lead to users having to write extra code to handle those cases.

when the other two options almost definitely lead to all kinds of user effort, in a less structured way.

A fourth option is... don't return the proxy for non-proxyable singletons, instead return the original object. They're (usually) going to be small. `None` is maybe a perfectly valid proxy for `None`: It's small to move around the network and behaves as if it was `None`, which is pretty much the goal for a proxy object, in the abstract.

**Comment by benclifford** [comment]
From the perspective of using ProxyStore as a potential serializer inside the existing serializer code which already chooses from a range of different serializers, I think that I would prefer option 1 (not the fourth option I mentioned above): that code already knows how to move on to a different serializer when an earlier serializer admits it cannot/does not want to perform a serialization.

**Comment by gpauloski** [comment]
My thoughts with option 1 are having to write something like:
```python
value = func_returns_obj_or_none()
if value is not None:
    value = store.proxy(value)
```
or
```python
value = func_returns_obj_or_singleton()
try:
    value = store.proxy(value)
except NonProxiableTypeError:
    pass
```

That being said, I do agree that trying to make it work when it clearly won't in some cases is unwise.

What do you think about having the default be to raise an error for the singleton types, but adding a `skip_unsupported_types: bool` flag that enables your proposed option 4?

**Comment by benclifford** [comment]
> What do you think about having the default be to raise an error for the singleton types, but adding a skip_unsupported_types: bool flag that enables your proposed option 4?

that seems fine.

although option 4 is the lightweight user-doesn't-have-to-think-too-hard path, so maybe you want that to be the default for lightweight users, and then people who really want option 1 style trouble can turn that on themselves. I don't have any serious preference there though - that's more about how you perceive people using this, i guess.

**Comment by gpauloski** [comment]
Thanks, @benclifford!

I fixed this in #315, and it'll be out in a v0.5.1 release sometime this week.

I ended up going with the default `skip_nonproxiable: bool = False` because (1) proxying None/True/False is unnecessary and not recommended and (2) it keeps the same default semantics of `Store.proxy()` that the return type is a `Proxy` (otherwise any existing code checked with MyPy would suddenly raise a bunch of errors).

MyPy should also correctly infer if the type is `Proxy` or a built-in constant when `skip_nonproxiable=True`. (Except for `Store.proxy_batch()` on a list of mixed types.)
